skimage/measure/segmentation_metrics.py

```diff
@@ -0,0 +1,315 @@
+    im_test_r = im_test.ravel()
+    im_true_r = im_true.ravel()
+    ignored = np.zeros(im_test_r.shape, np.bool)
+    data = np.ones(im_true_r.shape)
+    ignored[im_test_r == 0] = True
+    ignored[im_true_r == 0] = True
+    data[ignored] = 0
+    cont = sparse.coo_matrix((data, (im_test_r, im_true_r))).tocsr()
+    return cont
+
+def _xlogx(x):
+    """Compute x * log_2(x).
+
+    We define 0 * log_2(0) = 0
+
+    Parameters
+    ----------
+    x : ndarray or scipy.sparse.csc_matrix or csr_matrix
+        The input array.
+
+    Returns
+    -------
+    y : same type as x
+        Result of x * log_2(x).
+    """
+    y = x.copy()
+    if isinstance(y, sparse.csc_matrix) or isinstance(y, sparse.csr_matrix):
+        z = y.data
+    else:
+        z = np.asarray(y)  # ensure np.matrix converted to np.array
+    nz = z.nonzero()
+    z[nz] *= np.log2(z[nz])
+    return y
+
+def _divide_rows(matrix, column):
+    """Divide each row of `matrix` by the corresponding element in `column`.
+
+    The result is as follows: out[i, j] = matrix[i, j] / column[i]
+
+    Parameters
+    ----------
+    matrix : ndarray, scipy.sparse.csc_matrix or csr_matrix, shape (M, N)
+        The input matrix.
+    column : a 1D ndarray, shape (M,)
+        The column dividing `matrix`.
+
+    Returns
+    -------
+    out : same type as `matrix`
+        The result of the row-wise division.
+    """
+    out = matrix.copy()
+    if type(out) in [sparse.csc_matrix, sparse.csr_matrix]:
+        if type(out) == sparse.csr_matrix:
+            convert_to_csr = True
+            out = out.tocsc()
+        else:
+            convert_to_csr = False
+        column_repeated = np.take(column, out.indices)
+        nz = out.data.nonzero()
+        out.data[nz] /= column_repeated[nz]
+        if convert_to_csr:
+            out = out.tocsr()
+    else:
+        out /= column[:, np.newaxis]
+    return out
+
+def _divide_columns(matrix, row):
+    """Divide each column of `matrix` by the corresponding element in `row`.
+
+    The result is as follows: out[i, j] = matrix[i, j] / row[j]
+
+    Parameters
+    ----------
+    matrix : ndarray, scipy.sparse.csc_matrix or csr_matrix, shape (M, N)
+        The input matrix.
+    column : a 1D ndarray, shape (N,)
+        The row dividing `matrix`.
+
+    Returns
+    -------
+    out : same type as `matrix`
+        The result of the row-wise division.
+    """
+    out = matrix.copy()
+    if type(out) in [sparse.csc_matrix, sparse.csr_matrix]:
+        if type(out) == sparse.csc_matrix:
+            convert_to_csc = True
+            out = out.tocsr()
+        else:
+            convert_to_csc = False
+        row_repeated = np.take(row, out.indices)
+        nz = out.data.nonzero()
+        out.data[nz] /= row_repeated[nz]
+        if convert_to_csc:
+            out = out.tocsc()
+    else:
+        out /= row[np.newaxis, :]
+    return out
+
+def _vi_tables(im1, im2):
+    """Return probability tables used for calculating VI.
```

Comment:
I've made some improvements to the code in this function in Ch5 of Elegant SciPy. Can you copy over those improvements as well?